trucks.js

```diff
@@ -501,4 +513,173 @@ var foodTrucks = [
+  
+  trucks = foodTrucks.filter(function (currentTruck) {
+    if ((currentTruck.schedule) && (currentTruck.schedule.length > 0)) {
+      // Use the Array.prototype.find function to find the specified 'day',
+      // ignoring case, in the current truck's schedule.
+      dayMatched = currentTruck.schedule.find(function (currentDay) {
+        return dayRegExpr.test(currentDay);
+      });
+
+      // If there is a match, dayMatched contains the value of the
+      // matched element in the current truck's schedule array. Otherwise,
+      // dayMatched contains 'undefined'.
+      if (dayMatched) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+  });
+
+  return trucks;
+};
+
+/*************************************************************************
+* Description : filterTrucksByFoodType(foodType) - return trucks with
+*               associated 'foodType'
+* Parameters  : foodType - a string
+* Return      : Array of trucks that sell 'foodType'
+*************************************************************************/
+var filterTrucksByFoodType = function (foodType) {
+  var trucks = [];
+  var foodTypeMatched;
+  // Set foodTypeRegExpr to '/value of foodType/i' to search for
+  // 'value of foodType' ignoring case.
+  var foodTypeRegExpr = new RegExp(foodType, 'i');
+
+  trucks = foodTrucks.filter(function (currentTruck) {
+    if ((currentTruck.type) && (currentTruck.type.length > 0)) {
+      foodTypeMatched = currentTruck.type.find(function (currentFoodType) {
+        return foodTypeRegExpr.test(currentFoodType);
+      });
+    }
+    if (foodTypeMatched) {
+      return true;
+    } else {
+      return false;
+    }
+  });
+
+  return trucks;
+};
+
+/*************************************************************************
+* Description : addTruck(objTruck) - This method takes one parameter, a
+*               truck object, and adds it to the foodTrucks array.
+* Parameters  : objTruck - a truck object
+* Return      : the truck object if the operation is successful;
+*               otherwise, null is returned.
+*************************************************************************/
+var addTruck = function (objTruck) {
+  var result = null;
+
+  if ((objTruck !== null) && (typeof objTruck === 'object')) {
+    foodTrucks.push(objTruck);
+    result = objTruck;
+  }
+
+  return result;
+};
+
+/*************************************************************************
+* Description : removeTruck(name) - This method takes one parameter, a
+*               name associated with a food truck. It should remove the
+*               given truck
+* Parameters  : name - a string
+* Return      : 'true' if the operation is successful; 'false' otherwise
+*************************************************************************/
+var removeTruck = function (name) {
+  var result = false;
+  var truckToBeRemoved = getTruck(name);
+
+  if (truckToBeRemoved !== null) {
+    foodTrucks.pop(truckToBeRemoved);
```

Comment:
The `pop` method of an array will always remove the last item in the array (it doesn't take any parameters.) If you delete a truck in your list, it will appear to remove the one you just clicked the delete button for, because the client code removes the right one, but as soon as you refresh the page, you'll notice that one come back and the last item in the list will be gone. Instead, what you want to do is find the index of the relevant truck object in the foodTrucks array, and then use that index to call the array method `splice` to remove an item from an arbitrary spot in the array. Check out the documentation for the method on MDN if you have questions about it.